pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/OrDocIdIterator.java

```diff
@@ -15,139 +15,119 @@
  */
 package com.linkedin.pinot.core.operator.dociditerators;
 
-import java.util.Iterator;
-import java.util.PriorityQueue;
-import java.util.concurrent.atomic.AtomicLong;
-
-import com.linkedin.pinot.common.utils.Pairs;
-import com.linkedin.pinot.common.utils.Pairs.IntPair;
 import com.linkedin.pinot.core.common.BlockDocIdIterator;
 import com.linkedin.pinot.core.common.Constants;
 
-public final class OrDocIdIterator implements BlockDocIdIterator {
-  private BlockDocIdIterator[] docIdIterators;
-
-  final PriorityQueue<IntPair> queue;
-  final boolean[] iteratorIsInQueue;
-  int currentDocId = -1;
-  private int minDocId;
-  private int maxDocId;
-  private AtomicLong timeMeasure = new AtomicLong(0);
-  IntPair[] pointers;
 
-  /**
-   * @param docIdIterators
-   */
-  public OrDocIdIterator(BlockDocIdIterator[] docIdIterators) {
-    this.docIdIterators = docIdIterators;
-    queue =
-        new PriorityQueue<IntPair>(docIdIterators.length, new Pairs.AscendingIntPairComparator());
-    iteratorIsInQueue = new boolean[docIdIterators.length];
-    pointers= new IntPair[docIdIterators.length];
-    for (int i = 0; i < docIdIterators.length; i++) {
-      pointers[i] = new IntPair(0, i);
+public final class OrDocIdIterator implements BlockDocIdIterator {
+  private final BlockDocIdIterator[] _docIdIterators;
+  private final int[] _nextDocIds;
+  private final int _minDocId;
+  private final int _maxDocId;
+
+  private int _numNotExhaustedIterators;
+  private int _currentDocId = -1;
+
+  public OrDocIdIterator(BlockDocIdIterator[] docIdIterators, int minDocId, int maxDocId) {
+    _docIdIterators = docIdIterators;
+    _numNotExhaustedIterators = docIdIterators.length;
+    _nextDocIds = new int[_numNotExhaustedIterators];
+    for (int i = 0; i < _numNotExhaustedIterators; i++) {
+      _nextDocIds[i] = docIdIterators[i].advance(minDocId);
     }
+    _minDocId = minDocId;
+    _maxDocId = maxDocId;
+    removeExhaustedIterators();
   }
 
   @Override
-  public int advance(int targetDocId) {
-    if (currentDocId == Constants.EOF) {
+  public int next() {
```

Comment:
I got the logic by reading the code, it will be good to document the logic on top of the method.

Reply:
Added